server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/AbstractRecipientRewriteTableMailetTest.java

```diff
@@ -0,0 +1,347 @@
+
+        mailet.service(mail);
+
+        assertThat(mail.getMessage().getHeader(RFC2822Headers.RETURN_PATH)).containsOnly("<>");
+    }
+
+    @Test
+    public void serviceShouldResetValueOfMessageHeaderWhenNotNullSender() throws Exception {
+        mailet.init(mailetConfig);
+        mail = FakeMail.builder()
+            .sender(MailAddressFixture.ANY_AT_JAMES)
+            .mimeMessage(message)
+            .recipients(ImmutableList.of(MailAddressFixture.ANY_AT_JAMES, MailAddressFixture.OTHER_AT_JAMES))
+            .build();
+
+        mailet.service(mail);
+
+        assertThat(mail.getMessage().getHeader(RFC2822Headers.RETURN_PATH)).containsOnly("<" + MailAddressFixture.ANY_AT_JAMES + ">");
+    }
+
+    @Test
+    public void serviceShouldWriteLogAndSendMailToAllErrorRecipients() throws Exception {
+        mailetConfig = new FakeMailetConfig("Test", mockMailetContext);
+
+        mailet.init(mailetConfig);
+        mail = FakeMail.builder()
+            .mimeMessage(message)
+            .recipients(ImmutableList.of(MailAddressFixture.ANY_AT_JAMES, MailAddressFixture.OTHER_AT_JAMES))
+            .build();
+
+        mailet.service(mail);
+
+        verify(mockMailetContext, times(1)).log(eq(LogLevel.INFO), eq("Error while process mail."), any(MessagingException.class));
+        verify(mockMailetContext).sendMail(eq(mail.getSender()), eq(ImmutableList.of(MailAddressFixture.OTHER_AT_JAMES)), eq(message), eq(Mail.ERROR));
+        verifyNoMoreInteractions(mockMailetContext);
+    }
+
+    @Test
+    public void serviceShouldNotSendMailWhenNoErrorRecipients() throws Exception {
+        mailetConfig = new FakeMailetConfig("Test", mockMailetContext);
+
+        mailet.init(mailetConfig);
+        mail = FakeMail.builder()
+            .mimeMessage(message)
+            .recipients(ImmutableList.of(MailAddressFixture.ANY_AT_JAMES, nonDomainWithDefaultLocal))
+            .build();
+
+        mailet.service(mail);
+
+        verifyNoMoreInteractions(mockMailetContext);
+    }
+
+    @Test
+    public void serviceShouldResetMailStateToGhostWhenCanNotBuildNewRecipient() throws Exception {
+        mailet.init(mailetConfig);
+        mail = FakeMail.builder()
+            .mimeMessage(message)
+            .recipients(MailAddressFixture.OTHER_AT_JAMES, nonDomainWithDefaultLocal)
+            .build();
+
+        mailet.service(mail);
+
+        assertThat(mail.getState()).isEqualTo(Mail.GHOST);
+        assertThat(mail.getRecipients()).isEmpty();
+    }
+
+    @Test(expected = MessagingException.class)
+    public void handleMappingsShouldThrowExceptionWhenMappingsContainAtLeastOneNoneDomainObjectButCannotGetDefaultDomain() throws Exception {
+        when(domainList.getDefaultDomain()).thenThrow(DomainListException.class);
+        mailet.setDomainList(domainList);
+        mappings = MappingsImpl.fromRawString(MailAddressFixture.ANY_AT_JAMES.toString() + "," + NONEDOMAIN + "," + MailAddressFixture.OTHER_AT_JAMES.toString());
```

Comment:
- "MappingsImpl.fromRawString" is still there for a long time

Ok, sorry, I meant it was backported by Antoine. Of course it existed for a long time. But I don't think it is pertinant here. Sorry for this digression.

 - it's used to build mappings from string

That's the problem too ! Why do more parsing than needed ?

 - I think I will keep to raw string as parameter to make it more readable

Not in my opinion

```
mappings = MappingsImpl.fromRawString(MailAddressFixture.ANY_AT_JAMES.toString() + "," + NONEDOMAIN + "," + MailAddressFixture.OTHER_AT_JAMES.toString()); 
```

versus

```
mappings = MappingsImpl.builder()
    .add(MailAddressFixture.ANY_AT_JAMES.toString())
    .add(NONEDOMAIN)
    .add(MailAddressFixture.OTHER_AT_JAMES.toString())
    .build();
```

You could even improve MappingsImpl.Builder to makes it accept MailAddress addition, avoiding redoundant toString() call.

I definitly prefer the second.